File: python/剑指offer-51-数组中的逆序对.py
# hard

'''2020-10-19'''
class Solution:
    inverse_pair = 0
    def reversePairs(self, nums: List[int]) -> int:
        self.mergeSort(nums)
        return self.inverse_pair


    def mergeSort(self, alist):
        if len(alist) <= 1:
            return
    
        mid = len(alist) // 2
        lefthalf = alist[:mid]
        righthalf = alist[mid:]
        self.mergeSort(lefthalf)
        self.mergeSort(righthalf)
    
        i = j = k = 0
        while i < len(lefthalf) and j < len(righthalf):
            if lefthalf[i] <= righthalf[j]:
                alist[k] = lefthalf[i]
                i = i + 1
            else:
                alist[k] = righthalf[j]
                j = j + 1
                self.inverse_pair += (len(lefthalf) - i)
            k = k + 1
    
        while i < len(lefthalf):
            alist[k] = lefthalf[i]
            i = i + 1
            k = k + 1
    
        while j < len(righthalf):
            alist[k] = righthalf[j]
            j = j + 1
            k = k + 1

File: go/剑指offer-51-数组中的逆序对.go
// hard
// 2022-01-14
func reversePairs(nums []int) int {
    pairs := 0
    sortArray(nums, &pairs)
    return pairs
}

func sortArray(nums []int, pairs *int) []int {
    // MergeSort
    if len(nums) <= 1 {
        return nums
    }
    leftHalf := sortArray(nums[:len(nums)/2], pairs)
    rightHalf := sortArray(nums[len(nums)/2:], pairs)
    i, j, k := 0, 0, 0
    result := make([]int, len(nums))
    for i < len(leftHalf) && j < len(rightHalf) {
        if leftHalf[i] <= rightHalf[j] {
            result[k] = leftHalf[i]
            i++
            k++
        } else {
            result[k] = rightHalf[j]
            j++
            k++
            *pairs += len(leftHalf) - i
        }
    }
    if i < len(leftHalf) {
        for k < len(nums) {
            result[k] = leftHalf[i]
            i++
            k++
        }
    } else if j < len(rightHalf) {
        for k < len(nums) {
            result[k] = rightHalf[j]
            j++
            k++
        }
    }
    return result
}